Implemented Helpdesk

SanJSp opened the pull request:
This is an implementation of a Helpdesk using two dialogs to select a course and to further create a helpdesk ticket that is send to the api. 
The Helpdesk is accessible from the side drawer and from every activity via the options menu. 

First the Ticket creation dialog is opened. If it hasn't been called from inside a course, then a course/topic selection is needed. Otherwise the course will be auto selected. 

Next up the ticket has to be created. In the background information on the device and language will be transmitted parallel to the entered information (title, topic, issue).
An email is not needed if the user is already logged in.

Text fields that have been accessed but remained empty will raise an error.
If the ticket creation is canceled and information has been entered a cancelation confirmation dialog is opened.


Please review @tobiasrohloff.
Thank you.

![photo5350407358569884482](https://user-images.githubusercontent.com/38314662/68545054-961cc880-03c9-11ea-9961-e65aa22fc485.jpg)
![photo5350407358569884481](https://user-images.githubusercontent.com/38314662/68545053-961cc880-03c9-11ea-9a1b-6becd062eee3.jpg)
![photo5350624954498001949](https://user-images.githubusercontent.com/38314662/68545052-961cc880-03c9-11ea-9a9d-29713a5ce1d0.jpg)
![photo5350407358569884480](https://user-images.githubusercontent.com/38314662/68545050-95843200-03c9-11ea-97b5-a840b6ba8c39.jpg)

bengelhaupt commented on app/src/main/java/de/xikolo/controllers/cast/CastActivity.java:
```diff
@@ -24,14 +27,15 @@ public boolean onCreateOptionsMenu(Menu menu) {
         super.onCreateOptionsMenu(menu);
         getMenuInflater().inflate(R.menu.cast, menu);
         CastButtonFactory.setUpMediaRouteButton(this, menu, R.id.media_route_menu_item);
+        getMenuInflater().inflate(R.menu.helpdesk, menu);
```
The overflow menu in `CastActivity` has the wrong theme applied to it.
<img width="200px" src="https://user-images.githubusercontent.com/26904189/68666686-a6a57e00-0544-11ea-9d87-6b88704e7a26.jpg" />

bengelhaupt commented on app/src/main/java/de/xikolo/controllers/dialogs/ConfirmCancelDialog.kt:
```diff
@@ -0,0 +1,37 @@
+            .setNegativeButton(R.string.dialog_negative) { dialog: DialogInterface, _: Int ->
+                dialog.dismiss()
+            }
+            .setPositiveButton(getString(R.string.helpdesk_dialog_discard)) { dialog: DialogInterface, _: Int ->
```
The string resource should not be named specific to `helpdesk`. What about `dialog_discard_positive`?

SanJSp replied:
Done

bengelhaupt commented on app/src/main/java/de/xikolo/controllers/dialogs/ConfirmCancelDialog.kt:
```diff
@@ -0,0 +1,37 @@
+                listener?.onConfirmCancel()
+                dialog.dismiss()
+            }
+            .setMessage(getString(R.string.helpdesk_discard_dialog_text))
```
String again specific to `helpdesk`.

bengelhaupt commented on app/src/main/java/de/xikolo/controllers/dialogs/CreateTicketDialog.kt:
```diff
@@ -0,0 +1,192 @@
+
+        //listeners for wrong entries in EditTexts after focusing another view
+        titleEditText.setOnFocusChangeListener { _, b ->
+            if (titleEditText.text.isEmpty() && !b) titleEditText.error = getString(R.string.helpdesk_title_error)
```
Please use braces.

SanJSp replied:
Done

bengelhaupt commented on app/src/main/java/de/xikolo/controllers/dialogs/CreateTicketDialog.kt:
```diff
@@ -0,0 +1,192 @@
+        }
+
+        messageEditText.setOnFocusChangeListener { _, b ->
+            if (messageEditText.text.isEmpty() && !b) messageEditText.error = getString(R.string.helpdesk_message_error)
```
Braces

SanJSp replied:
Done

bengelhaupt commented on app/src/main/java/de/xikolo/controllers/dialogs/CreateTicketDialog.kt:
```diff
@@ -0,0 +1,192 @@
+        }
+
+        emailEditText.setOnFocusChangeListener { _, b ->
+            if ((!Patterns.EMAIL_ADDRESS.matcher(emailEditText.text).matches() || emailEditText.text.isEmpty()) && emailEditText.visibility != View.GONE && !b) emailEditText.error = getString(R.string.helpdesk_email_error)
```
Braces

SanJSp replied:
Done

bengelhaupt commented on app/src/main/java/de/xikolo/models/dao/CourseDao.kt:
```diff
@@ -17,9 +17,17 @@ class CourseDao(realm: Realm) : BaseDao<Course>(Course::class, realm) {
     override fun find(id: String?): LiveData<Course> =
         query()
-            .beginGroup()
+                .beginGroup()
```
`.beginGroup()` should not be on the same indentation level as `equalTo()` or `or()`

bengelhaupt commented on app/src/main/java/de/xikolo/models/dao/CourseDao.kt:
```diff
@@ -34,7 +42,7 @@ class CourseDao(realm: Realm) : BaseDao<Course>(Course::class, realm) {
                 Realm.getDefaultInstance().use { realm ->
                     realm.where<Course>()
-                        .beginGroup()
+                            .beginGroup()
```
indentation level

bengelhaupt commented on app/src/main/java/de/xikolo/models/dao/CourseDao.kt:
```diff
@@ -150,9 +158,9 @@ class CourseDao(realm: Realm) : BaseDao<Course>(Course::class, realm) {
-                            .greaterThanOrEqualTo("endDate", Date())
+                        .isNull("endDate")
+                        .or()
+                        .greaterThanOrEqualTo("endDate", Date())
```
Needs to be indented one tab

bengelhaupt commented on app/src/main/java/de/xikolo/models/dao/CourseDao.kt:
```diff
@@ -176,9 +184,9 @@ class CourseDao(realm: Realm) : BaseDao<Course>(Course::class, realm) {
-                            .lessThan("endDate", Date())
+                        .isNotNull("endDate")
+                        .and()
+                        .lessThan("endDate", Date())
```
Indentation

bengelhaupt commented on app/src/main/java/de/xikolo/viewmodels/helpdesk/TicketViewModel.kt:
```diff
@@ -0,0 +1,27 @@
+    private val coursesDao = CourseDao(realm)
+    private val courseListDelegate = CourseListDelegate(realm)
+
+    val courses = coursesDao.all()
```
Can be replaced by `val courses = courseListDelegate.courses`. This removes the `CourseDao` dependency.

bengelhaupt commented on app/src/main/res/values-de/strings.xml:
```diff
@@ -295,7 +296,24 @@
+    <string name="helpdesk_title_error">Bitte Titel eingeben</string>
+    <string name="helpdesk_email_error">Bitte gültige E-Mail Adresse eingeben</string>
+    <string name="helpdesk_dialog_discard">Verwerfen</string>
+    <string name="helpdesk_discard_dialog_text">Ticket verwerfen?</string>
```
See comments above about resource naming

bengelhaupt commented on app/src/main/res/values/strings.xml:
```diff
@@ -328,4 +324,23 @@
+    <string name="helpdesk_message_error">Please describe your issue</string>
+    <string name="helpdesk_email_error">Please enter a valid e-mail</string>
+    <string name="helpdesk_dialog_discard">Discard</string>
+    <string name="helpdesk_discard_dialog_text">Discard ticket?</string>
```
Resource naming

bengelhaupt commented on build.gradle:
```diff
@@ -13,7 +13,7 @@ buildscript {
 
         xikoloApiVersion    = 3
-        realmSchemaVersion  = 8
+        realmSchemaVersion  = 7
```
I checked again and `realmSchemaVersion` should actually be `8` like on `master`.

bengelhaupt commented on app/src/main/res/values/strings.xml:
```diff
@@ -328,4 +324,23 @@
     <string name="empty_message_course_certificates">No certificates available for this course</string>
     <string name="empty_message_dates">No upcoming dates right now</string>
 
+    <string name="settings_helpdesk">Help</string>
```
unused resource

bengelhaupt commented on app/src/main/res/values/strings.xml:
```diff
@@ -328,4 +324,23 @@
+    <string name="helpdesk_dialog_send">Send</string>
+    <string name="helpdesk_default_topic">Please choose a topic</string>
+    <string name="helpdesk_title_error">Please enter a title</string>
+    <string name="helpdesk_message_error">Please describe your issue</string>
```
missing german version

bengelhaupt commented on app/src/main/res/values/strings.xml:
```diff
@@ -324,23 +324,21 @@
     <string name="helpdesk_info_text">Please describe your technical problem using the textboxes below. Information about your mobile device and your e-mail address will be automatically submitted along with your issue report.</string>
-    <string name="helpdesk_title_hint">Title</string>
-    <string name="helpdesk_topic_hint">Topic</string>
-    <string name="helpdesk_mail_hint">Email</string>
```
what happened to the mail hint?

SanJSp replied:
Reused Ressource from Line 148
<string name="hint_email">Email</string>

bengelhaupt replied:
But then it is still included in the german resources

SanJSp replied:
true, deleted

tobiasrohloff commented on app/src/main/res/layout/dialog_helpdesk_create_ticket.xml:
```diff
@@ -0,0 +1,111 @@
+            android:layout_width="0dp"
+            android:layout_height="wrap_content"
+            android:layout_marginStart="24dp"
+            android:layout_marginTop="16dp"
```
@bengelhaupt have you already moved dimens like these to variables as well?

bengelhaupt replied:
No, dimen consolidation is still a thing that needs to be done.

tobiasrohloff commented on app/src/main/res/values-de/strings.xml:
```diff
@@ -191,7 +192,7 @@
     <string name="course_date_type_course_start">Kursstart</string>
     <string name="course_date_type_section_start">Sektionsstart</string>
     <string name="course_date_type_submission_publishing">Aufgabe</string>
-    <string name="course_date_type_submission_deadline">Abgabefrist</string> <!-- ToDo -->
```
What about this change? What was the todo about?

SanJSp replied:
Finding a better translation

SanJSp replied:
@bengelhaupt I remember that this was one of your todos that I could remove. Is this correct?

bengelhaupt replied:
It was a todo annotation of me I forgot to remove. It was about translation if I remember correctly.

tobiasrohloff commented on app/src/main/res/values-de/strings.xml:
```diff
@@ -295,7 +296,25 @@
+    <string name="helpdesk_default_topic">Bitte wählen Sie ein Thema</string>
+    <string name="helpdesk_title_error">Bitte Titel eingeben</string>
+    <string name="helpdesk_message_error">Bitte Anliegen beschreiben</string>
+    <string name="helpdesk_email_error">Bitte gültige E-Mail Adresse eingeben</string>
```
Different spellings of `E-Mail`. Please align with other locals.

tobiasrohloff commented on app/src/main/res/values-de/strings.xml:
```diff
@@ -295,7 +296,25 @@
+    <string name="helpdesk_message_error">Bitte Anliegen beschreiben</string>
+    <string name="helpdesk_email_error">Bitte gültige E-Mail Adresse eingeben</string>
+    <string name="dialog_discard_positive">Verwerfen</string>
+    <string name="dialog_discard_inputs">Eingaben verwerfen?</string>
```
`Möchten Sie alle Ihre Eingaben verwerfen?`

tobiasrohloff commented on app/src/main/res/values-de/strings.xml:
```diff
@@ -295,7 +296,25 @@
+    <!-- -->
+    <string name="helpdesk_topic_technical_question">Technische Frage</string>
+    <string name="hint_issue">Anliegen</string>
+    <string name="helpdesk_info_text">Bitte beschreiben Sie Ihr Anliegen mit Hilfe des Formulars unten. Informationen über Ihr Mobilgerät und Ihre E-Mail-Adresse werden automatisch mit Ihrer Anfrage übermittelt.</string>
```
First part only `Bitte beschreiben Sie Ihr Anliegen.`

Automatic email input is not always guaranteed. Text should adjust based on case.

tobiasrohloff commented on app/src/main/res/values/colors.xml:
```diff
@@ -20,6 +20,7 @@
 
 
     <!-- Content backgrounds -->
+    <color name="bg">#FFFFFF</color>
```
Maybe that name should be more specific? What kind of bg is that? Where is it used?

SanJSp replied:
set background color to transparent

tobiasrohloff commented on app/src/main/res/values/strings.xml:
```diff
@@ -328,4 +324,21 @@
 
+    <string name="helpdesk_topic_technical_question">Technical question</string>
+    <string name="hint_issue">Issue</string>
+    <string name="helpdesk_info_text">Please describe your technical problem using the textboxes below. Information about your mobile device and your e-mail address will be automatically submitted along with your issue report.</string>
```
Same feedback as for German.

tobiasrohloff commented on app/src/main/res/values/strings.xml:
```diff
@@ -328,4 +324,21 @@
+    <string name="helpdesk_message_error">Please describe your issue</string>
+    <string name="helpdesk_email_error">Please enter a valid e-mail</string>
+    <string name="dialog_discard_positive">Discard</string>
+    <string name="dialog_discard_inputs">Discard input?</string>
```
`Do you want to discard all of your input?`

SanJSp commented on app/src/main/java/de/xikolo/controllers/announcement/AnnouncementActivity.kt:
```diff
@@ -51,6 +55,21 @@ class AnnouncementActivity : CollapsingToolbarActivity() {
+        when (item.itemId) {
+            R.id.action_helpdesk -> {
+                val dialog = CreateTicketDialogAutoBundle.builder().build()
+                dialog.show(supportFragmentManager, CreateTicketDialog.TAG)
```
In other classes mith multiple items it has been done like I did (CourseActivity), why should I change it?

SanJSp commented on app/src/main/java/de/xikolo/models/dao/CourseDao.kt:
```diff
@@ -195,7 +203,6 @@ class CourseDao(realm: Realm) : BaseDao<Course>(Course::class, realm) {
                         .findAll()
                         .asCopy()
                 }
-
```
No, autoformatiing removed it. I added one like it was before my editing.

SanJSp commented on app/src/main/java/de/xikolo/controllers/base/BaseMetaRecyclerViewAdapter.kt:
```diff
@@ -65,9 +65,6 @@ abstract class BaseMetaRecyclerViewAdapter<M, S> : RecyclerView.Adapter<Recycler
     class HeaderViewHolder(view: View) : RecyclerView.ViewHolder(view) {
 
-        @BindView(R.id.container)
-        lateinit var container: ViewGroup
```
Otherwise the app crashes with a recyclerView Ressource not found error

SanJSp commented on app/src/main/java/de/xikolo/controllers/base/BaseMetaRecyclerViewAdapter.kt:
```diff
@@ -47,7 +47,7 @@ abstract class BaseMetaRecyclerViewAdapter<M, S> : RecyclerView.Adapter<Recycler
     }
 
-    protected fun createHeaderViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
+    protected open fun createHeaderViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
```
jep, in HelpDeskTopicAdapter

SanJSp commented on app/src/main/java/de/xikolo/controllers/dialogs/CreateTicketDialog.kt:
```diff
@@ -0,0 +1,198 @@
+
+    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+        return AlertDialog.Builder(activity!!, R.style.AppTheme_Dialog)
+            .setNegativeButton(R.string.dialog_negative) { _: DialogInterface, _: Int ->
```
well you have to pass something, android studio proposed to replace the placeholders with underscores

SanJSp commented on app/src/main/java/de/xikolo/controllers/dialogs/HelpdeskTopicDialog.kt:
```diff
@@ -0,0 +1,126 @@
+        return calendar.get(Calendar.YEAR)
+    }
+
+    private fun overlappingClose() {
```
renamed to closeDialog

SanJSp commented:
In the latest commit I've implemented a device offline handling. When your device is offline and you're trying to open the dialog, it will not open and show a toast. Furthermore this test will happen again immeadiately before sending the ticket, which prevents sending of tickets without a connection.

Added to that the ChooseTopicDialog was changed. Instead of the numbers at the beginning of the title, I've implemented a different style which includes the course slug, cours year and the course title by itself.

![photo5402195228116495478](https://user-images.githubusercontent.com/38314662/69743173-4b77ac00-113e-11ea-9982-501c50b48704.jpg)
![photo5402195228116495479](https://user-images.githubusercontent.com/38314662/69743174-4b77ac00-113e-11ea-9944-0987cd51dfac.jpg)

bengelhaupt commented on app/src/main/java/de/xikolo/controllers/dialogs/CreateTicketDialog.kt:
```diff
@@ -57,6 +59,12 @@ class CreateTicketDialog : ViewModelDialogFragment<TicketViewModel>(), HelpdeskT
 
+        if (!context.isOnline) {
+            showToast(R.string.toast_no_network)
+            dialog!!.cancel()
```
Can we make it safe aka `dialog?.cancel()`?

bengelhaupt commented on app/src/main/java/de/xikolo/controllers/dialogs/CreateTicketDialog.kt:
```diff
@@ -124,18 +132,28 @@ class CreateTicketDialog : ViewModelDialogFragment<TicketViewModel>(), HelpdeskT
+            getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener {
+                if (context.isOnline) {
+                    viewModel.send(titleEditText.text.toString(), messageEditText.text.toString(), topic, emailEditText.text.toString(), courseId)
+                    (dialog as AlertDialog).dismiss()
```
Is it always safe to cast the dialog to an `AlertDialog`? Otherwise a safe cast `(dialog as? AlertDialog)?.dismiss()` can be used.

bengelhaupt replied:
Actually it does not need to be casted at all in this case but the question still applies for line 131.

SanJSp replied:
It is not always called as AlertDialog therefore I will implement the safe cast

bengelhaupt commented on app/src/main/java/de/xikolo/controllers/dialogs/CreateTicketDialog.kt:
```diff
@@ -124,18 +132,28 @@ class CreateTicketDialog : ViewModelDialogFragment<TicketViewModel>(), HelpdeskT
 
     override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+
+
```
Empty lines

bengelhaupt commented on app/src/main/java/de/xikolo/controllers/dialogs/CreateTicketDialog.kt:
```diff
@@ -124,18 +132,28 @@ class CreateTicketDialog : ViewModelDialogFragment<TicketViewModel>(), HelpdeskT
             }
             .setPositiveButton(R.string.dialog_send) { _: DialogInterface, _: Int ->
-                viewModel.send(titleEditText.text.toString(), messageEditText.text.toString(), topic, emailEditText.text.toString(), courseId)
+
```
Why are those lambdas empty?

bengelhaupt commented on app/src/main/java/de/xikolo/controllers/dialogs/HelpdeskTopicAdapter.kt:
```diff
@@ -59,11 +71,21 @@ class HelpdeskTopicAdapter(private val onTopicClickedListener: OnTopicClickedLis
     }
 
     private fun bindTopicViewHolder(holder: CourseTitleViewHolder, position: Int) {
+
```
empty line

bengelhaupt commented on app/src/main/java/de/xikolo/controllers/dialogs/HelpdeskTopicAdapter.kt:
```diff
@@ -59,11 +71,21 @@ class HelpdeskTopicAdapter(private val onTopicClickedListener: OnTopicClickedLis
 
         holder.layout.setOnClickListener {
-            onTopicClickedListener?.onTopicClicked(topic.first, topic.second, topic.third)
+            onTopicClickedListener?.onTopicClicked(holder.topicTitle.text.toString(), topic.first, topic.second)
```
Can we introduce a local variable for this first argument? I can only think of side effects when using the `TextView` getter.
Like
```
var t = if(topic.first == TicketTopic.COURSE){
    ...
    course?.title
} else {
    ...
    topic.first.toString()
}
holder.topicTitle.text = t
...
onTopicClickedListener(t,...)

bengelhaupt commented on app/src/main/java/de/xikolo/controllers/dialogs/HelpdeskTopicAdapter.kt:
```diff
@@ -79,4 +101,10 @@ class HelpdeskTopicAdapter(private val onTopicClickedListener: OnTopicClickedLis
+        calendar.time = course!!.startDate
+        return calendar.get(Calendar.YEAR)
+    }
+
```
This would fit perfectly inside `DateExtensions`. It could then be used like `course.startDate.year`

bengelhaupt commented on app/src/main/res/layout/item_helpdesk_topic_list.xml:
```diff
@@ -1,29 +1,81 @@
 <?xml version="1.0" encoding="utf-8"?>
 <androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
```
Debugging leftovers

bengelhaupt commented on app/src/main/res/layout/item_helpdesk_topic_list.xml:
```diff
@@ -1,29 +1,81 @@
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toTopOf="parent"
+            tools:text="2019" />
```
Debugging leftovers
